File: app/src/main/java/com/proxy/shadowsocksr/impl/UDPEncryptor.java
package com.proxy.shadowsocksr.impl;

import com.proxy.shadowsocksr.impl.crypto.CryptoManager;
import com.proxy.shadowsocksr.impl.crypto.crypto.AbsCrypto;

import java.util.Arrays;

public final class UDPEncryptor
{
    private final AbsCrypto crypto;

    private final int[] cryptMethodInfo;//Key size and iv size.

    public UDPEncryptor(String pwd, String cryptMethod)
    {
        cryptMethodInfo = CryptoManager.Companion.getCipherInfo(cryptMethod);
        crypto = CryptoManager.Companion.getMatchCrypto(cryptMethod, pwd);
    }

    public int getIVLen()
    {
        return cryptMethodInfo[1];
    }

    public byte[] encrypt(byte[] buf)
    {
        byte[] iv = Utils.Companion.srandomBytes(cryptMethodInfo[1]);
        crypto.updateEncryptIV(iv);
        buf = crypto.encrypt(buf);
        byte[] data = new byte[iv.length + buf.length];
        System.arraycopy(iv, 0, data, 0, iv.length);
        System.arraycopy(buf, 0, data, iv.length, buf.length);
        return data;
    }

    public byte[] decrypt(byte[] buf)
    {
        byte[] iv = Arrays.copyOfRange(buf, 0, cryptMethodInfo[1]);
        buf = Arrays.copyOfRange(buf, cryptMethodInfo[1], buf.length);
        crypto.updateDecryptIV(iv);
        return crypto.decrypt(buf);
    }
}

File: jvmVer/src/main/java/com/proxy/shadowsocksr/impl/UDPEncryptor.kt
package com.proxy.shadowsocksr.impl

import com.proxy.shadowsocksr.impl.crypto.CryptoManager
import com.proxy.shadowsocksr.impl.crypto.crypto.AbsCrypto

import java.util.Arrays

class UDPEncryptor(pwd: String, cryptMethod: String)
{
    private val crypto: AbsCrypto

    private val cryptMethodInfo: IntArray//Key size and iv size.

    init
    {
        cryptMethodInfo = CryptoManager.getCipherInfo(cryptMethod)
        crypto = CryptoManager.getMatchCrypto(cryptMethod, pwd)
    }

    val ivLen: Int get() = cryptMethodInfo[1]

    fun encrypt(buf: ByteArray): ByteArray
    {
        var bf = buf
        val iv = ImplUtils.srandomBytes(cryptMethodInfo[1])
        crypto.updateEncryptIV(iv)
        bf = crypto.encrypt(bf)
        val data = ByteArray(iv.size + bf.size)
        System.arraycopy(iv, 0, data, 0, iv.size)
        System.arraycopy(bf, 0, data, iv.size, bf.size)
        return data
    }

    fun decrypt(buf: ByteArray): ByteArray
    {
        var bf = buf
        val iv = Arrays.copyOfRange(bf, 0, cryptMethodInfo[1])
        bf = Arrays.copyOfRange(bf, cryptMethodInfo[1], bf.size)
        crypto.updateDecryptIV(iv)
        return crypto.decrypt(bf)
    }
}
